### PR TITLE
Feature/PM-1049 - TDE Flow 3 -  login decryption options component

### PR DESCRIPTION
**Pull request by JaredSnider-Bitwarden**
## Type of change

```
- [ ] Bug fix
- [X] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Code changes

- **file.ext:** Description of what was changed and why

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

### CODE REVIEW

File: libs/angular/src/auth/components/base-login-decryption-options.component.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        device.type === DeviceType.AndroidAmazon ||
+        device.type === DeviceType.WindowsDesktop ||
+        device.type === DeviceType.MacOsDesktop ||
+        device.type === DeviceType.LinuxDesktop ||
+        device.type === DeviceType.UWP
+      ) {
+        this.showApproveFromOtherDeviceBtn = true;
+        break;
+      }
+    }
```
Comment (coroiu) [inline libs/angular/src/auth/components/base-login-decryption-options.component.ts:58]: suggestion: consider converting this into an observable 

```
this.approveFromOtherDeviceEnabled$ = from(this.devicesApiService.getDevices())
  .pipe(map(response) => {
    const devices = response.data;
    return devices.some(device.type === DeviceType.Android || /* ... */);
   })
```

**Review by coroiu — COMMENTED** [review]
Some initial comments after working with the code a little bit

File: apps/desktop/src/auth/login/login-decryption-options/login-decryption-options.component.html [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        <span id="rememberThisDeviceHintText">{{ "uncheckIfPublicDevice" | i18n }}</span>
+      </form>
+
+      <div class="buttons with-rows">
+        <div class="buttons-row" *ngIf="showApproveFromOtherDeviceBtn">
+          <button
+            (click)="approveFromOtherDevice()"
+            type="button"
+            class="btn primary block"
+            [appA11yTitle]="'approveFromYourOtherDevice' | i18n"
```
Comment (coroiu) [inline apps/desktop/src/auth/login/login-decryption-options/login-decryption-options.component.html:34]: suggestion: I think we can remove `appA11yTitle` when it's identical to the displayed content

File: apps/browser/src/auth/popup/login-decryption-options/login-decryption-options.component.html [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        >
+          {{ "approveWithMasterPassword" | i18n }}
+        </button>
+      </div>
+
+      <hr class="muted-hr mx-5px mb-20px" />
+
+      <div class="small mx-5px">
+        <p class="no-margin">{{ "loggingInAs" | i18n }} {{ userEmail }}</p>
+        <a [routerLink]="[]" (click)="logOut()">{{ "notYou" | i18n }}</a>
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply apps/browser/src/auth/popup/login-decryption-options/login-decryption-options.component.html:73]: Good question. I copied the existing code from the [web login component](https://github.com/bitwarden/clients/blob/4124f7bdc8753ffc666f18b0099760f3e95d9976/apps/web/src/app/auth/login/login.component.html#LL131C11-L131C11). 

Hmm... it appears that the router link was simply added to get the benefit of the href styling (blue text + underline + tab focus). The click event will fire first and send the logout message via the messagingService, and then the router will redirect to the root of the app... If the logout takes any distinguishable amount of time, then the user gets a bad experience in being navigated to the root of the app and then the login page. Unfortunately, there isn't a promise that I can await to complete for the logout process. 

However, I can get around the issue by replacing the use of the routerLink with a few added attributes:
```
   <a tabindex="0" role="button" style="cursor: pointer" (click)="logOut()">{{
          "notYou" | i18n
   }}</a>
```

Figma:
image

Real HTML:
image

However, technically, this is a logout action -- it doesn't really match the existing "not you" logic on the login page as the user isn't authenticated in that context. It is much more similar to the lock screen logout button:
image
I wonder if I should ask design if I should instead match the existing log out button implementation...

File: libs/angular/src/auth/components/base-login-decryption-options.component.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    // Client will create an AuthRequest of type AdminAuthRequest WITHOUT orgId and send it to the server
+    // Server will look up the org id(s) based on the user id and create the AdminAuthRequest(s)
+    // Note: must lookup if the user has an account recovery key (resetPasswordKey) set in the org
+    // (means they've opted into the Admin Acct Recovery feature)
+    // Per discussion with Micah, fire out requests to all admins in any orgs the user is a member of
+    // UNTIL the Admin Console team finishes their work to turn on Single Org policy when Admin Acct Recovery is enabled.
+  }
+
+  approveWithMasterPassword() {
+    this.router.navigate(["/lock"]);
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/angular/src/auth/components/base-login-decryption-options.component.ts:102]: I normally would not, but when I initially created the stub methods, I thought that we might need to perform additional logic before navigation. I think we will need to save the user's choice for whether or not they are trusting the device to a service for use later on after decryption in order to establish trust once we have the user's symmetric key.

File: libs/angular/src/auth/components/base-login-decryption-options.component.ts (hunk 1 above)
Comment (JaredSnider-Bitwarden) [reply libs/angular/src/auth/components/base-login-decryption-options.component.ts:58]: @coroiu , I've refactored the base component to leverage async piped observable streams across the board. 
My draft PR for the changes is [here](https://github.com/bitwarden/clients/pull/5642).

File: libs/angular/src/auth/components/base-login-decryption-options.component.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+    protected router: Router,
+    protected messagingService: MessagingService,
+    protected tokenService: TokenService,
+    protected loginService: LoginService
+  ) {}
+
+  async ngOnInit() {
+    // Determine if the user has any mobile or desktop devices
+    // to determine if we should show the approve from other device button
+    const devicesListResponse = await this.devicesApiService.getDevices();
```
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply libs/angular/src/auth/components/base-login-decryption-options.component.ts:44]: This has now been resolved in https://github.com/bitwarden/clients/pull/5642 by the addition of an observable based device data store service which sits in front of the API service and will position itself well for the future SDK ownership of the API services.

File: apps/browser/src/auth/popup/login-decryption-options/login-decryption-options.component.html (hunk 3 above)
[the remark this answers is not in the document]
Comment (JaredSnider-Bitwarden) [reply apps/browser/src/auth/popup/login-decryption-options/login-decryption-options.component.html:73]: Agreed re CL in all clients. `<button bitLink [linkType]="linkType">Button</button>` would be perfect here. 

I attempted to use a button there, but the styling is a pain. I'm going to keep with the link without routerLink for now as it meets the design requirements and is consistent w/ the existing browser client implementation (even though it probably should say "log out")

**Pull request closed without merging**